packages/shared/scope-hoisting/src/generate.js

```diff
@@ -1,44 +1,78 @@
 // @flow
 
-import type {Bundle, BundleGraph} from '@parcel/types';
-import type {File} from '@babel/types';
+import type {Bundle, BundleGraph, PluginOptions} from '@parcel/types';
+import type {
+  ExpressionStatement,
+  File,
+  Statement,
+  StringLiteral,
+} from '@babel/types';
 
 import babelGenerate from '@babel/generator';
 import nullthrows from 'nullthrows';
 import {isEntry} from './utils';
+import SourceMap from '@parcel/source-map';
+import * as t from '@babel/types';
+import template from '@babel/template';
 
-export function generate(bundleGraph: BundleGraph, bundle: Bundle, ast: File) {
-  let {code} = babelGenerate(ast, {
-    minified: bundle.env.minify,
-    comments: true, // retain /*@__PURE__*/ comments for terser
-  });
+const REGISTER_TEMPLATE = template.statement<
+  {|ID: StringLiteral, STATEMENTS: Array<Statement>|},
+  ExpressionStatement,
+>('parcelRequire.registerBundle(ID, function () { STATEMENTS; })');
+const WRAPPER_TEMPLATE = template.statement<
+  {|STATEMENTS: Array<Statement>|},
+  ExpressionStatement,
+>('(function () { STATEMENTS; })()');
 
+export function generate(
+  bundleGraph: BundleGraph,
+  bundle: Bundle,
+  ast: File,
+  options: PluginOptions,
+) {
   // $FlowFixMe
   let interpreter: ?string = bundle.target.env.isBrowser()
     ? null
     : nullthrows(bundle.getMainEntry()).meta.interpreter;
-  let hashBang = interpreter != null ? `#!${interpreter}\n` : '';
 
   let entry = bundle.getMainEntry();
   let isAsync = entry && !isEntry(bundle, bundleGraph);
-  if (!bundle.env.minify && (isAsync || bundle.env.outputFormat === 'global')) {
-    code = `\n${code}`;
-  }
 
   // Wrap async bundles in a closure and register with parcelRequire so they are executed
   // at the right time (after other bundle dependencies are loaded).
-  let contents = '';
+  let statements = ast.program.body;
   if (isAsync && bundle.env.outputFormat === 'global') {
-    contents = `${hashBang}parcelRequire.registerBundle(${JSON.stringify(
-      nullthrows(entry).id,
-    )},function(){${code}\n});`;
-  } else {
-    contents =
-      hashBang +
-      (bundle.env.outputFormat === 'global'
-        ? `(function(){${code}\n})();`
-        : code);
+    statements = [
+      REGISTER_TEMPLATE({
+        ID: t.stringLiteral(nullthrows(entry).id),
+        STATEMENTS: statements,
+      }),
+    ];
+  } else if (bundle.env.outputFormat === 'global') {
+    statements = [WRAPPER_TEMPLATE({STATEMENTS: statements})];
   }
 
-  return {contents};
+  ast = t.file(
+    t.program(
+      // $FlowFixMe Statement is incompatible with BabelNodeStatement
```

Comment:
That fixme doesn't actually seem to be needed?

Reply:
It is 😞

Reply:
Well, I can remove the fixme and flow doesn't complain.
(Statement and BabelNodeStatement are two `type`s to the same sum type).

Reply:
Strange — you're right. I remember this being an error before. @padmaia can attest to this I think 😅

Reply:
I also remember flow warning about unnecessary fixme's,  but not anymore...